dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,245 @@
+     * @deprecated prefer {@link #pinUntilError}
+     */
+    @Deprecated
+    public static PinUntilErrorChannel pinUntilErrorWithoutReshuffle(List<LimitedChannel> channels) {
+        return new PinUntilErrorChannel(channels, ThreadLocalRandom.current());
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        if (channels.isEmpty()) {
+            log.debug("Rejecting request due to no delegates");
+            return Optional.empty();
+        }
+
+        if (reshuffle.isPresent()) {
+            reshuffle.get().reshuffleChannelsIfNecessary();
+        }
+
+        int currentIndex = currentHost.get();
+        LimitedChannel channel = channels.get(currentIndex % channels.size());
+
+        Optional<ListenableFuture<Response>> maybeFuture = channel.maybeExecute(endpoint, request);
+        if (!maybeFuture.isPresent()) {
+            OptionalInt next = incrementCurrentHost(currentIndex);
+            debugLogCurrentChannelRejected(currentIndex, channel, next);
+            return Optional.empty(); // if the caller retries immediately, we'll get the next host
+        }
+
+        ListenableFuture<Response> future = maybeFuture.get();
+        return Optional.of(DialogueFutures.addDirectCallback(future, new FutureCallback<Response>() {
+            @Override
+            public void onSuccess(Response response) {
+                if (response.code() / 100 <= 2) {
+                    // we consider all the 1xx and 2xx codes successful, so want to remain pinned to this channel
+                    return;
+                }
+
+                OptionalInt next = incrementCurrentHost(currentIndex);
+                debugLogReceivedErrorStatus(currentIndex, channel, response, next);
+
+                // TODO(dfox): handle 308 See Other somehow, as we currently don't have a host -> channel mapping
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                OptionalInt next = incrementCurrentHost(currentIndex);
+                debugLogReceivedThrowable(currentIndex, channel, throwable, next);
+            }
+        }));
+    }
+
+    /**
+     * If we have some reason to think the currentIndex is bad, we want to move to the next host. This is done with a
+     * compareAndSet to ensure that out of order responses which signal information about a previous host don't kick
+     * us off a good one.
+     */
+    private OptionalInt incrementCurrentHost(int currentIndex) {
+        int next = currentIndex + 1;
+        int nextIndex = (next >= 0) ? next % channels.size() : 0; // we don't want negative indexes!
+
+        boolean saved = currentHost.compareAndSet(currentIndex, nextIndex);
+        return saved ? OptionalInt.of(nextIndex) : OptionalInt.empty(); // we've moved on already
+    }
+
+    private final class Reshuffle {
```

Comment:
would it be cleaner to provide a view over the list of channels that handles reshuffling on access thus isolating the reshuffle logic

Reply:
Just pushed a version where it's hidden behind a single abstraction, what do you think?